File: Practica 3/practica3.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

// Prototipos de funciones
double suma(double a, double b);
double resta(double a, double b);
double multiplicacion(double a, double b);
double division(double a, double b);
double potencia(double a, double b);
double raiz_cuadrada(double a);
double factorial(double a);
double grados_a_radianes(double grados);
double seno(double a);
double coseno(double a);
double tangente(double a);
int menu();

// Funciones de operaciones matemáticas
double suma(double a, double b) {
    return a + b;
}

double resta(double a, double b) {
    return a - b;
}

double multiplicacion(double a, double b) {
    return a * b;
}

double division(double a, double b) {
    if (b != 0) {
        return a / b;
    } else {
        printf("Error: No se puede dividir por cero.\n");
        exit(EXIT_FAILURE);
    }
}

double potencia(double a, double b) {
    return pow(a, b);
}

double raiz_cuadrada(double a) {
    if (a >= 0) {
        return sqrt(a);
    } else {
        printf("Error: No se puede calcular la raiz cuadrada de un numero negativo.\n");
        exit(EXIT_FAILURE);
    }
}

double factorial(double a) {
    if (a >= 0) {
        if (a == 0) {
            return 1;
        } else {
            return a * factorial(a - 1);
        }
    } else {
        printf("Error: No se puede calcular el factorial de un numero negativo.\n");
        exit(EXIT_FAILURE);
    }
}
double grados_a_radianes(double grados) {
    return grados * M_PI / 180.0;
}
double seno(double a) {
    return sin(a);
}

double coseno(double a) {
    return cos(a);
}

double tangente(double a) {
    return tan(a);
}

// Función para mostrar el menú y obtener la opción del usuario
int menu() {
    int opcion;
    printf("\nSeleccione una operacion:\n");
    printf("1. Suma\n");
    printf("2. Resta\n");
    printf("3. Multiplicacion\n");
    printf("4. Division\n");
    printf("5. Potencia\n");
    printf("6. Raiz cuadrada\n");
    printf("7. Factorial\n");
    printf("8. Seno\n");
    printf("9. Coseno\n");
    printf("10. Tangente\n");
    printf("11. Salir\n");
    printf("Ingrese el numero de la operacion que desea realizar: ");
    scanf("%d", &opcion);
    return opcion;
}

// Función principal
int main() {
    while (1) {
        int opcion = menu();

        if (opcion == 11) {
            printf("Hasta luego\n");
            break;
        }

        double a, b;
        printf("Ingrese el primer numero: ");
        scanf("%lf", &a);
        
        if (opcion != 6 && opcion != 7 && opcion != 8 && opcion != 9 && opcion != 10) {
            printf("Ingrese el segundo numero: ");
            scanf("%lf", &b);
        }

        switch(opcion) {
            case 1:
                printf("Resultado: %.2lf\n", suma(a, b));
                break;
            case 2:
                printf("Resultado: %.2lf\n", resta(a, b));
                break;
            case 3:
                printf("Resultado: %.2lf\n", multiplicacion(a, b));
                break;
            case 4:
                printf("Resultado: %.2lf\n", division(a, b));
                break;
            case 5:
                printf("Resultado: %.2lf\n", potencia(a, b));
                break;
            case 6:
                printf("Resultado: %.2lf\n", raiz_cuadrada(a));
                break;
            case 7:
                printf("Resultado: %.2lf\n", factorial(a));
                break;
            case 8:
                printf("Resultado: %.2lf\n", seno(grados_a_radianes(a)));
                break;
            case 9:
                printf("Resultado: %.2lf\n", coseno(grados_a_radianes(a)));
                break;
            case 10:
                printf("Resultado: %.2lf\n", tangente(grados_a_radianes(a)));
                break;
            default:
                printf("Opcion invalida. Intente nuevamente.\n");
        }
    }
    return 0;
}

File: Practica 3/practica3.py
import math

# Definir las funciones para las operaciones aritméticas
def suma(a, b):
    return a + b
def resta(a, b):
    return a - b
def multiplicacion(a, b):
    return a * b
def division(a, b):
    if b == 0:
        return "Error: No se puede dividir por cero."
    else:
        return a / b
        
# Funciones para otras operaciones
def raiz_cuadrada(a):
    if a < 0:
        return "Error: No se puede calcular la raíz cuadrada de un número negativo."
    else:
        return math.sqrt(a)
def raiz_n(a, n):
    if a < 0 and n % 2 == 0:
        return "Error: No se puede calcular la raíz par de un número negativo."
    else:
        return a ** (1/n)
def exponencial(a, b):
    return a ** b
def seno(a):
    return math.sin(a)
def coseno(a):
    return math.cos(a)
def tangente(a):
    return math.tan(a)
def arcoseno(a):
    return math.asin(a)
def arcocoseno(a):
    return math.acos(a)
def arcotangente(a):
    return math.atan(a)

# Función principal
def calculadora():
    operaciones = {
        "suma": suma,
        "resta": resta,
        "multiplicacion": multiplicacion,
        "division": division,
        "raiz_cuadrada": raiz_cuadrada,
        "raiz_n": raiz_n,
        "exponencial": exponencial,
        "seno": seno,
        "coseno": coseno,
        "tangente": tangente,
        "arcseno": arcoseno,
        "arccoseno": arcocoseno,
        "arctangente": arcotangente
    }
    while True:
        try:
            operacion = input("\nIngrese la operación (suma, resta, multiplicacion, division, raiz_cuadrada, raiz_n, exponencial, seno, coseno, tangente, arcseno, arccoseno, arctangente): ")
            if operacion in operaciones:
                if operacion == "raiz_cuadrada":
                    num = float(input("Ingrese el número para calcular la raíz cuadrada: "))
                    resultado = operaciones[operacion](num)
                elif operacion == "raiz_n":
                    num = float(input("Ingrese el número: "))
                    n = int(input("Ingrese el índice de la raíz: "))
                    resultado = operaciones[operacion](num, n)
                elif operacion in ["seno", "coseno", "tangente", "arcseno", "arccoseno", "arctangente"]:
                    num = float(input("Ingrese el ángulo en radianes: "))
                    resultado = operaciones[operacion](num)
                else:
                    num1 = float(input("\nIngrese el primer número: "))
                    num2 = float(input("Ingrese el segundo número: "))
                    resultado = operaciones[operacion](num1, num2)
                print("El resultado es:", resultado)
            else:
                print("Operación no válida. Intente nuevamente.")
        except ValueError:
            print("Entrada inválida. Intente nuevamente.")
        continuar = input("\n¿Desea realizar otra operacion? (s/n): ")
        if continuar.lower() != "s":
            break
# Llamar a la función principal
calculadora()